src/qe/mbp/mbp_term_graph.cpp

```diff
@@ -445,130 +740,355 @@ namespace mbp {
             return expr_ref(res, m);
         }
 
-        res = mk_app_core (r.get_expr());
+        res = mk_app_core(r.get_expr());
         m_term2app.insert(r.get_id(), res);
         return expr_ref(res, m);
-
     }
 
     expr_ref term_graph::mk_app(expr *a) {
-        term *t = get_term(a);
-        if (!t)
-            return expr_ref(a, m);
-        else
-            return mk_app(t->get_root());
-
+      term *t = get_term(a);
+      if (!t)
+        return expr_ref(a, m);
+      else {
+        SASSERT(t->get_repr());
+        return mk_app(*t->get_repr());
+      }
     }
 
-    void term_graph::mk_equalities(term const &t, expr_ref_vector &out) {
-        SASSERT(t.is_root());
+    void term_graph::mk_equalities(term &t, expr_ref_vector &out) {
+        SASSERT(t.is_repr());
+        if(t.get_class_size() == 1) return;
         expr_ref rep(mk_app(t), m);
         for (term *it = &t.get_next(); it != &t; it = &it->get_next()) {
             expr* mem = mk_app_core(it->get_expr());
             out.push_back (m.mk_eq (rep, mem));
         }
     }
 
-    void term_graph::mk_all_equalities(term const &t, expr_ref_vector &out) {
+
+    void term_graph::mk_all_equalities(term &t, expr_ref_vector &out) {
+        if (t.get_class_size() == 1)
+            return;
+
         mk_equalities(t, out);
 
         for (term *it = &t.get_next(); it != &t; it = &it->get_next ()) {
-            expr* a1 = mk_app_core (it->get_expr());
+            expr* a1 = mk_app_core(it->get_expr());
             for (term *it2 = &it->get_next(); it2 != &t; it2 = &it2->get_next()) {
-                expr* a2 =  mk_app_core(it2->get_expr());
-                out.push_back (m.mk_eq (a1, a2));
+              expr *a2 = mk_app_core(it2->get_expr());
+              out.push_back(m.mk_eq(a1, a2));
             }
         }
     }
 
+    void term_graph::mk_qe_lite_equalities(term &t, expr_ref_vector &out, filter_core* in_core) {
+        SASSERT(t.is_repr());
+        if (t.get_class_size() == 1) return;
+        expr_ref rep(m);
+        rep = mk_app(t);
+        if (in_core && !(*in_core)(rep)) return;
```

Comment:
Should `!(*in_core)(rep)` be in an assertion? Are there cases where the rep is in the core?